scvi/model/_condscvi.py

```diff
@@ -121,41 +119,88 @@ def get_vamp_prior(
+            y = tensors[REGISTRY_KEYS.LABELS_KEY]
+            out = self.module.inference(x, y)
+            mean_, var_ = out["qz_m"], out["qz_v"]
+            mean += [mean_.cpu()]
+            var += [var_.cpu()]
+
+        mean_cat, var_cat = np.array(torch.cat(mean)), np.array(torch.cat(var))
+        adata.obsm["X_CondSCVI"] = mean_cat
+
+        for ct in range(self.summary_stats["n_labels"]):
+            local_indices = np.where(adata.obs[key] == mapping[ct])[0]
+            if "overclustering_vamp" not in adata.obs.columns:
+                if p < len(local_indices) and p > 0:
+                    from sklearn.cluster import KMeans
+
+                    kmeans = KMeans(n_clusters=p, n_init=30).fit(
+                        mean_cat[local_indices]
+                    )
+                    overclustering_vamp = kmeans.labels_
+                else:
+                    # Every cell is its own cluster
+                    overclustering_vamp = list(range(len(local_indices)))
+            else:
+                overclustering_vamp = adata[local_indices, :].obs["overclustering_vamp"]
+
+            keys, counts = np.unique(overclustering_vamp, return_counts=True)
+
+            n_labels_overclustering = len(keys)
+            if n_labels_overclustering > p:
+                raise ValueError(
+                    "Given overclustering contains more clusters than vamp_prior_p. Increase value."
+                )
+            var_cluster = np.zeros(
+                [
+                    n_labels_overclustering,
+                    self.module.n_latent,
+                ]
             )
-            mean = []
-            var = []
-            for tensors in scdl:
-                x = tensors[REGISTRY_KEYS.X_KEY]
-                y = tensors[REGISTRY_KEYS.LABELS_KEY]
-                out = self.module.inference(x, y)
-                mean_, var_ = out["qz_m"], out["qz_v"]
-                mean += [mean_.cpu()]
-                var += [var_.cpu()]
-
-            mean_vprior[ct], var_vprior[ct] = np.array(torch.cat(mean)), np.array(
-                torch.cat(var)
+            mean_cluster = np.zeros(
+                [
+                    n_labels_overclustering,
+                    self.module.n_latent,
+                ]
             )
 
-        return mean_vprior, var_vprior
+            for index, cluster in enumerate(np.unique(overclustering_vamp)):
+                indices_curr = local_indices[
+                    np.where(overclustering_vamp == cluster)[0]
+                ]
+                var_cluster[index, :] = (
+                    np.mean(var_cat[indices_curr], axis=0)
+                    + np.mean(mean_cat[indices_curr] ** 2, axis=0)
+                    - (np.mean(mean_cat[indices_curr], axis=0)) ** 2
+                )
+                mean_cluster[index, :] = np.mean(mean_cat[indices_curr], axis=0)
+
+            mean_vprior[ct, 0:n_labels_overclustering, :] = mean_cluster
```

Comment:
you can use a slice object so this slice code is not repeated in the following three lines